import kgo-verifier

nvartolomei opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [x] v25.1.x
- [x] v24.3.x
- [ ] v24.2.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

jcsp committed:
Initial verifier with random read

jcsp committed:
Try GC explicity after each consume

jcsp committed:
Fix consumer to pick the right partition

jcsp committed:
Fix consume code

jcsp committed:
Sequential read, bad offsets

jcsp committed:
WIP

jcsp committed:
Raise error on partial getOffsets results

jcsp committed:
Track successfully written ranges rather than errror offsets

jcsp committed:
Fix valid ranges write

jcsp committed:
Add --parallel option

jcsp committed:
Fill out readme, log memory stats

jcsp committed:
simplify randomRead

jcsp committed:
Fix readme formatting

jcsp committed:
strip memory usage debugging

The franz-go leak still exists, but it's narrowed
down to the actual consume code so no need
to keep spamming about it everywhere.

jcsp committed:
Disable SASL if no username given

jcsp committed:
Update to latest franz-go

jcsp committed:
Update franz-go to pick up memory leak fix

This enables using much higher numbers of
iteratinos in --rand_read_msgs

jcsp committed:
Update readme for memory leak fix

jcsp committed:
randomRead: clean up logging & add timeout

* We now log errors from the PollRecords return value
* Add a tag to the log message from each parallel worker
* Log if PollRecords returns no records
* Pass a timeout context into PollRecords

jcsp committed:
Rename to kgo-verifier

jcsp committed:
Add a license

jcsp committed:
Write produce status messages to stdout

jcsp committed:
Output incremental status for reads

jcsp committed:
Tweak random reads to only count non-empty reads

...when deciding when to drop out of the loop.  This
gives a deterministic result for the ValidReads statistic,
rather than having that statistic depend on how many
empty reads we experience.

jcsp committed:
Output status on time intervals, not msg counts

This makes tests more robust when using
walltime timeouts

jcsp committed:
Write offset files atomically, initialize checkpoint timestamps

jcsp committed:
.github: add a dependabot config

NyaliaLui committed:
main: add batchMaxBytes command line arg

ztlpn committed:
move ConsumerStatus funcs closer to the type definition

jcsp committed:
Merge pull request #1 from NyaliaLui/add-batchmaxbytes-option

main: add batchMaxBytes command line arg

ztlpn committed:
add consumer group validating reader

jcsp committed:
Merge pull request #2 from ztlpn/add-consumer-group-reader

Add consumer group reader

jcsp committed:
Move main.go into cmd/ path

jcsp committed:
Add kgo-repeater executable

This is a companion to verifier, which does not attempt
offset verification but does an easy-to-scale traffic
generation loop where a population of clients use a
consumer group to split up the partitions, and then
each client re-produces messages it consumes, to keep
a steady flow of traffic through a system under test.

ballard26 committed:
main: log error for produce failures

jcsp committed:
Merge pull request #3 from ballard26/additional-produce-log

Log error for produce failures

jcsp committed:
Update README for kgo-repeater

Commit:
Add a simple Dockerfile

jcsp committed:
Merge pull request #5 from VladLazar/add-dockerfile

Add a simple Dockerfile

jcsp committed:
kgo-repeater: make MaxBufferedRecords configurable

jcsp committed:
Merge pull request #4 from jcsp/wip-repeater

Add kgo-repeater traffic generator

jcsp committed:
kgo-verifier: refactor into worker classes

This is a precursor to generic status reads over HTTP.

jcsp committed:
kgo-verifier: Add HTTP status+stats endpoints

jcsp committed:
kgo-verifier: optionally stay alive until told to shut down

...so that tests can read the final status of e.g. a producer
when it finishes writing out all its messages.

jcsp committed:
kgo-verifier: do looping on consumers

This replaces a bash loop in tests.  To enable controlling
it, there is a new last_pass HTTP endpoint to ask it to
stop after it's current read, and a new 'active' element
in the status object to let the test know when this
worker has dropped out and its final status may be read.

jcsp committed:
kgo-verifier: improved error handling

jcsp committed:
kgo-verifier: simplify main

Don't try to do things like optionally running
seq consumer in parallel with random consumers: that's
not how the tools is used in the tests.

jcsp committed:
kgo-repeater: fix initialization

jcsp committed:
kgo-repeater: structured worker names + better logging

Give clients helpful names that simplify debugging
when a particular worker has an issue.  The name shows
up in:
- kgo-repeater -debug log
- kgo-repeater -trace log (franz-go logs)
- redpanda server as the client ID

Related: https://github.com/redpanda-data/redpanda/issues/5959

jcsp committed:
kgo-repeater: enable CPU profiling

jcsp committed:
kgo-repeater: switch to binary encoding

...from JSON, for the content of messages.

This is a >2x speedup, and moves the CPU
bottleneck from the workload generator
code into the franz-go code (specifically
sink.createReq is where it spends most of
its time).

jcsp committed:
Update franz-go to tip of master

This is to pick up two fixes for issues that were
found in ManyPartitionsTest:
 - https://github.com/twmb/franz-go/issues/198
 - https://github.com/twmb/franz-go/issues/190

The #198 symptom of clients dropping out of consumer
groups has been observed even after fixes, it is unclear if
there is another bug in franz-go or if something else
is going on.

Related: https://github.com/redpanda-data/redpanda/issues/5959

jcsp committed:
Merge pull request #8 from jcsp/updates-20220902

kgo-repeater logging+performance improvements, franz-go update for bug fixes

mmaslankaprv committed:
seq_read: log sequential read offsets with context

Signed-off-by: Michal Maslanka <michal@vectorized.io>

jcsp committed:
Merge pull request #9 from redpanda-data/fix-logging-seq-read-offsets

seq_read: log sequential read offsets with context

mmaslankaprv committed:
verifier: added name flag to set worker kafka client id

Signed-off-by: Michal Maslanka <michal@vectorized.io>

mmaslankaprv committed:
Merge pull request #10 from redpanda-data/added-name-flag

verifier: added name flag to set worker kafka client id

jcsp committed:
kgo-verifier: remove spurious status print

This status object was constructed and then logged,
producing confusing zero-message status lines in
the output.

jcsp committed:
kgo-verifier: log status instead of printing to stdout

This gets us timestamps on the output lines.  The stdout
printing was a hangover from when these lines used to
be how ducktape services received the status.

jcsp committed:
kgo-verifier: improve producer logging on unexpected offsets

- The "writing partition..." message printed one offset too high
- Fix a snake-case variable name
- If OffsetRanges calls Die, it should say what the acceptable range was
- producerInner should log its "Wrote partition" message before calling
  into validOffsetse.insert, so that we see the message even if that
  insert call dies.

jcsp committed:
kgo-verifier: make seq_read_worker more verbose

Historically, I kept these logs very terse for the benefit
of interactive use and because in earlier ducktape code
we took the stdout/stderr of the process over an SSH connection
back to the test runner.

Now that the services keep local log files that are hauled
back at the end of a test, we can afford to be much more verbose.

We still don't log the vector of offsets at info level on every poll call,
because that would be a huge log for ManyPartitionsTest.  Tests should
tune the logging up/down as needed: e.g. tests that operate on smaller
partition counts & smaller volumes of data can safely enable debug
logging.

mmaslankaprv committed:
Merge pull request #11 from jcsp/verifier-logging

kgo-verifier: improve logging to help debug apparent hangs

jcsp committed:
kgo-verifier: add a producer flag for faking timestamps

This is for automated testing of time queries in Redpanda.

jcsp committed:
Merge pull request #12 from jcsp/fake-timestamps

kgo-verifier: add a producer flag for faking timestamps

ballard26 committed:
kgo-verifier: add support for verifying transactions

ballard26 committed:
kgo-repeater: add support for transactions

jcsp committed:
Merge pull request #13 from ballard26/transactional-producer

Add support for verifying transactions

jcsp committed:
kgo-verifier: explicitly close client in seq_consumer

This avoids potentially leaving consumers
running until they get GC'd.  They should not be
actively doing anything, but their consumer loops
may still be spamming fetch RPCs to redpanda.

jcsp committed:
update franz-go to 1.9.1

ztlpn committed:
verifier: add seq_read_msgs option (max number of msgs to consume)

ztlpn committed:
verifier: add warning if there is no offsets file

ztlpn committed:
verifier add consumer-side rate limiting

ztlpn committed:
verifier: restore ability to stop by Ctrl-C in interactive mode

ztlpn committed:
verifier: fix printing last status for group worker

ztlpn committed:
verifier: add looping support for group read worker

jcsp committed:
Merge pull request #14 from jcsp/seq-consumer-close

kgo-verifier: explicitly close client in seq_consumer

ztlpn committed:
verifier: add random component to group consumer group name

Otherwise in the loop mode (if the loop duration is less than 1 second)
the group name can repeat and the consumer can freeze.

jcsp committed:
Merge pull request #15 from ztlpn/add-consume-throttling

Add consume throttling and consumed message count limit

jcsp committed:
kgo-verifier: fix transactional producer valid offset tracking

This was previously generating a fragmented valid_offsets file
that skipped over the transaction marker offsets.

jcsp committed:
kgo-verifier: fix transactional reads in seq_read_worker

jcsp committed:
kgo-verifier: include a count of aborted messages in status

This enables more robust tests: they can check this value
to calculate exactly how many messages a consumer should
expect to see.

ballard26 committed:
Merge pull request #17 from jcsp/tx-improvements

kgo-verifier: improve transaction verification

r-vasquez committed:
verifier: add timeout query param to /last_pass

This will create a go routine which will print
the stack trace after that timeout is completed.

We want to catch bugs that makes the program hang
with this we can debug the stack trace after the
given timeout

twmb committed:
Merge pull request #19 from r-vasquez/debug-stack-trace

verifier: add timeout query param to /last_pass

jcsp committed:
kgo-verifier: close client in produceInner

On "unexpected offsets" conditions, the producer
will drop out of produceInner and re-enter, to
destroy and create a client.

However, dropping the reference to the client doesn't
synchronously close it, so it's possible for the
client to remain alive too long, and perhaps continue
to drain its produce channel, disrupting the offset checks
in the subsequent call to produceInner.

It's not certain that we've seen this happen, but it is one
possibility discussed in https://github.com/redpanda-data/redpanda/issues/8289
that is straightforward to eliminate.

Related: https://github.com/redpanda-data/redpanda/issues/8289

jcsp committed:
kgo-verifier: add -produce-throughput-bps flag

Exactly the same as -consume-throughput-mb, but
for producers, and uses bytes per second to
enable using low throughputs for non-stress tests.

jcsp committed:
Merge pull request #20 from redpanda-data/sync-close-producer

kgo-verifier: close client in produceInner

jcsp committed:
Merge pull request #21 from redpanda-data/produce-rate-limit

kgo-verifier: add -produce-throughput-bps flag

jcsp committed:
kgo-repeater: add -rate-limit-bps

jcsp committed:
Fix stack trace print + add explicit /print_stack endpoint

This was previously only printing the stack of the current
goroutine.

The /print_stack endpoint enables tests to trigger a stack
dump when they can detet that something has gone wrong other
than a hang, such as when a consumer mysteriously drops
out of a consumer group.

jcsp committed:
Merge pull request #23 from jcsp/stack-dumps

Fix stack trace print + add explicit /print_stack endpoint

jcsp committed:
Merge pull request #22 from jcsp/repeater-rate-limit

kgo-repeater: add -rate-limit-bps

jcsp committed:
Update franz-go from 1.9.1 to 1.12.1

The previous release was ~3mo old

jcsp committed:
Merge pull request #24 from redpanda-data/update-franz-go

Update franz-go from 1.9.1 to 1.12.1

jcsp committed:
kgo-verifier: handle out of range offsets in random consumer

Related: https://github.com/redpanda-data/redpanda/issues/9423

jcsp committed:
Merge pull request #27 from redpanda-data/issue-9423-random-out-of-range

kgo-verifier: handle out of range offsets in random consumer

jcsp committed:
Add --compression-type, --compressible-payload flags

These are useful for stress testing compacted topics, where
Redpanda must decompress batches using the producer's choice of
codec, and incur whatever cpu/memory overhead that implies.

jcsp committed:
Remove unused function

Commit:
validator: use record headers to store record id

Previously verifier used key as a record identifier. To make record key
independent from the identifier moved the identifier to record header.
Headers are orthogonal to record key and value hence they do not
influence compaction and partitioning.

Signed-off-by: Michal Maslanka <michal@redpadna.com>

Commit:
added support for generating workloads with keys that are compactible

Added a parameter allowing user to request keys from a set of a
requested cardinality.

Signed-off-by: Michal Maslanka <michal@redpadna.com>

mmaslankaprv committed:
Merge pull request #29 from mmaslankaprv/compacted-topics-support

Compacted topics support

jcsp committed:
Merge pull request #28 from redpanda-data/compression-config

Configurable compression

jcsp committed:
kgo-repeater: don't include rate limiting in latency stat

jcsp committed:
Merge pull request #30 from redpanda-data/repeater-latency-rate-limit

kgo-repeater: don't include rate limiting in latency stat

jcsp committed:
Use debug=2 for pprof goroutine dumps

This should get us all the goroutines, not just
a subset.

jcsp committed:
Merge pull request #31 from jcsp/pprof-debug-2

Use debug=2 for pprof goroutine dumps

Commit:
kgo-verifier: track max consumed offset

This commit extends the status returned by the kgo-verifier workers
to contain the high watermarks for each partition (i.e. the max offset
consumed/produced).

A number of our tests that use the kgo-verifier consumer check
the number of valid reads in an attempt to ensure that all produced
offsets have been consumed at least once. Since the consumer follows a
sea-saw pattern (go back to the start offset when the end is reached),
this check is not necessarily correct. The new fields introduced in the
status can be used to improve those tests.

michael-redpanda committed:
kgo-verifier: Enabled support for TLS

Utilized the TLS dialer to enable TLS support for kgo-verifier.
Caveat: Must use a 'known' CA to sign certificates, or add your CA to
the OS's list of good CAs.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
Merge pull request #34 from michael-redpanda/enable-tls-in-kgo-verifier

kgo-verifier: Enabled support for TLS

michael-redpanda committed:
kgo-repeater: Added support for TLS and SASL

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
Merge pull request #35 from michael-redpanda/enable-tls-in-kgo-repeater

kgo-repeater: Added support for TLS and SASL

michael-redpanda committed:
worker/repeater: Remove override of compression

ProducerBatchCompression is provided by the worker config but was being
overridden to NoCompression.  Remove that line

Signed-off-by: Michael Boquard <michael@redpanda.com>

jcsp committed:
Merge pull request #36 from michael-redpanda/fix-kgo-repeater-compressio

worker/repeater: Remove override of compression

michael-redpanda committed:
status: Added topic name to status

Signed-off-by: Michael Boquard <michael@redpanda.com>

jcsp committed:
Merge pull request #37 from michael-redpanda/add-more-status

status: Added topic name to status

jcsp committed:
Merge pull request #33 from VladLazar/report-max-consumed-offset

kgo-verifier: track max consumed offset

Commit:
kgo-verifier: return status alongside lock

e02458e added a map to the various status structs returned by workers.
This turned out to be problematic as JSON serialisation could race with
mutation of the struct. This patch changes the GetStatus method to
return a tuple made out of the status and its lock. This approach was
taken because the status types differ and their locks are in different
places.

Commit:
kgo-verifier: add configurable timestamp step

Retention tests would benefit from spacing out their timestamps more, so
this patch introduces a new flag for the producer to control the
timestamp step. It's only used when fake timestamps are enabled.

I've also done a little refactor and removed the fakeTimestampMs from
the producer worker. The value from the config struct is used instead.

Commit:
Merge pull request #38 from VladLazar/fix-status-map-race

kgo-verifier: fix status map race and make timestamp step configurable

NyaliaLui committed:
kgo-verifier: update franzgo and its deps

Signed-off-by: NyaliaLui <nyalia@redpanda.com>

NyaliaLui committed:
Merge pull request #39 from NyaliaLui/update-franzgo-and-its-deps

kgo-verifier: update franzgo and its deps

NyaliaLui committed:
kgo-verifier: require go 1.19

Discovered that this requires 1.19 to build now. So the mod file should
indicate that.

Signed-off-by: NyaliaLui <nyalia@redpanda.com>

NyaliaLui committed:
Merge pull request #40 from NyaliaLui/requires-1.19

kgo-verifier: require go 1.19

Commit:
producer: added ability to recreate producer after given number of msgs

Added ability to recreate a producer client every specified number of
messages. Recreation of producer forces creation of new unique producer
id. This setup allow to stress Redpanda producer id cache implementation
as it generates lots of unique producer ids.

Signed-off-by: Michal Maslanka <michal@redpadna.com>

mmaslankaprv committed:
Merge pull request #41 from mmaslankaprv/msgs-per-producer

producer: added ability to recreate producer after given number of msgs

nvartolomei committed:
panic if the listener for remote control can not be started

Due to test infra issues we've had these services started where previous
instances were already running leading to very hard to debug results.

Panic right away instead of silently ignoring the error.

I'm choosing to do it in the most simple way as a graceful server
shutdown is not required for the use-cases on these services.

Before:

```
 → ./kgo-verifier -topic aba -debug -produce_msgs 10
INFO[0000] Getting topic metadata...
DEBU[0000] Targeting topic aba with 1 partitions
INFO[0000] Starting producer...
...
DEBU[0000] Writing partition 0 at 50
...
INFO[0000] Finished producer.
```

After:

```
 → ./kgo-verifier -topic aba -debug -produce_msgs 10
INFO[0000] Getting topic metadata...
DEBU[0000] Targeting topic aba with 1 partitions
INFO[0000] Starting producer...
INFO[0000] Loading offsets for topic aba t=-1...
panic: failed to listen on 0.0.0.0:7884: listen tcp 0.0.0.0:7884: bind: address already in use
```

twmb committed:
go.{mod,sum}: bump franz-go to close redpanda-data/redpanda#13791

See twmb/franz-go/commit/c013050

bharathv committed:
Merge pull request #43 from redpanda-data/franz

go.{mod,sum}: bump franz-go to close redpanda-data/redpanda#13791

nvartolomei committed:
Merge pull request #42 from redpanda-data/nv/err-on-listen-fail

panic if the listener for remote control can not be started

nvartolomei committed:
kgo-verifier: fix use-transactions flag description

nvartolomei committed:
kgo-verifier: panic on out of order reads

nvartolomei committed:
random_read_worker: fix offset formatting

offset is of kgo.Offset type. This is a struct rather than a number.

Using %s directive to get the result of the String() method

nvartolomei committed:
seq_read_worker: track and update lwm instead of last read

Previous implementation led to reading messages twice on errors as we
were using last_read as low watermark when resuming. Off-by-one error.

There was another bug which is fixed now. If PollFetches returned only
errors then sequentialReadInner would return an array with all offsets
set to 0, thus causing a read from beginning.

nvartolomei committed:
Merge pull request #45 from redpanda-data/nv/monotonic-kafka-offests

kgo-verifier: assert kafka offset monotonicity

Commit:
kgo-repeater: Remove unused topic metadata logic

Commit:
kgo-repeater: Pass topics list input to repeater

Commit:
kgo-repeater: Make program multi-topic aware

- Random topic from input is chosen as topic to produce to each produce
loop iteration

- Consumers will consume from all topics passed via input instead of
just one

michael-redpanda committed:
Merge pull request #47 from redpanda-data/repeater-multi-topic

Make the kgo-repeater multi-topic aware

nvartolomei committed:
kgo: run one extra pass after /last_pass request

This is being invoked by stress tests which run consumers in a loop but
want to ensure at the end that all the data was read.

kgo first requests high topic watermark, then starts routines that read
up to that watermark. If the last run begins before all the data is
written, then `/last_pass` is called, it will exit without having a
chance to read all the data.

mfleming committed:
go.{mod,sum}: Move to franz-go 1.15.4 to fix redpanda-data/redpanda#15131

mfleming committed:
Merge pull request #49 from mfleming/bump-franz-go

go.{mod,sum}: Move to franz-go 1.15.4 to fix redpanda-data/redpanda#15131

nvartolomei committed:
Merge pull request #48 from redpanda-data/nv/last-pass-inclusive

kgo: run one extra pass after /last_pass request

mmaslankaprv committed:
verifier: fixed stopping consumers when loop is requested

Previously the code didn't wait for `/last_pass` request before stopping
the consumer.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
Merge pull request #51 from mmaslankaprv/fix-loop-flag

verifier: fixed stopping consumers when loop is requested

nvartolomei committed:
verifier: allow to tolerate data loss

This mode allows to verify redpanda when write caching is enabled. In
addition to tolerating data loss we also record and export to the
/status endpoint the number of offsets/records that are considered lost
from the point of view of the verifier.

nvartolomei committed:
fix formatting directives typos

This fails `go test` which is introduced in subsequent commit.

nvartolomei committed:
pkg: introduce loop state object for managing consumer loops

Will be used for refactoring in subsequent PR.

nvartolomei committed:
verifier: use loop state object for managing consumer loops

nvartolomei committed:
.gitignore: vscode

nvartolomei committed:
verifier: continuous mode

In this mode verifier waits for new data to be produced instead of
exiting on EOF.

Stops after /last_pass request.

nvartolomei committed:
verifier: do not block on channels

Never hang /last_pass, /shutdown requests in case of retries.

Quality of life change.

nvartolomei committed:
verifier: allow custom consumer group names

This allows to restart the consumer and have it resume from the last
committed offset.

nvartolomei committed:
verifier: allow launching only a single type of consumer per instance

Will make state management easier in a subsequent PR.

nvartolomei committed:
.github: test in ci

nvartolomei committed:
Merge pull request #52 from redpanda-data/nv/write-caching

verifier: tolerate data loss

nvartolomei committed:
repeater: remove redundant consumer group option

This is set few lines before. Setting an empty string is no-op. No need
to check if it non-empty.

nvartolomei committed:
repeater: remove dead code

Leftovers from previous refactor /shrug

nvartolomei committed:
repeater: panic if group is not set

Repeater does not work without a consumer group. It will start but not
do any work.

nvartolomei committed:
Merge pull request #53 from redpanda-data/nv/repeater-validate-group-name

repeater: validate group name

nvartolomei committed:
verifier: do not crash with index out of range for invalid data

kgo expects to consume messages written by kgo itself. These messages
usually contain data (in headers) which is used for validation.

If unrelated data is produced to a partition it crashes because the
header it expects does not exist.

This commit avoids the crash and categorizes either categorizes the read
as out of scope or crashes if it should have been a valid read (running
with a kgo produced valid_offsets json).

nvartolomei committed:
Merge pull request #55 from redpanda-data/nv/do-not-crash-no-header

verifier: do not crash with index out of range for invalid data

nvartolomei committed:
crash instantly on data loss errors

We (nvartolomei and bash) found that GroupReadWorker wouldn't report
data loss. The reason is that we didn't fail the test instantly when
data loss is detected instead we relied on monotonicity validation to
fail. I.e. we would crash when we detected that we would consume an
earlier offset compared to what we consumed before.

In the GroupReadWorker we reset the monotonicity validation state on any
errors because the next attempt at consuming is almost certain to read
the second time offsets already consumed. As a result we wouldn't catch
monotonicity issues.

In retrospect, it would have been better to fail instantly when franz-go
detects data loss. So this commit achieves exactly that.

bashtanov committed:
producer: add an option to tolerate failures, count them

nvartolomei committed:
assert monotonic leader epochs

nvartolomei committed:
always validate and record consumed offsets

Previously we would record offsets only when consuming expected/valid
records.  This is overly restrictive as the offset validation and now
term validation is validating log semantics (monotonicity).

bashtanov committed:
consumer: combine offset map from multiple partially successful replies

If the number of partitions is high and leadership transfers happen often we may
be unable to get a single all-successful result for a long time. Instead,
combine a full result out of tolerate partially successful ones.

nvartolomei committed:
Merge pull request #58 from redpanda-data/nv/monotonic-terms

record all records and validate leader epoch monotonicity

bashtanov committed:
Merge pull request #57 from bashtanov/tolerate-failed-produce

producer: add an option to tolerate failures, count them

bashtanov committed:
Merge pull request #59 from bashtanov/combine-offset-results

consumer: combine offset map from multiple partially successful replies

WillemKauf committed:
`worker`: add `tombstone_probability` as parameter

Allows the `producer_worker` and `repeater_worker` to generate tombstone
records.

WillemKauf committed:
`worker`: add `compacted` parameter

Offset gaps exist in compacted topics. Suppress a warning log
in the read workers if the user passes a `--compacted` flag
as a parameter.

WillemKauf committed:
`worker`: add `TombstonesProduced` and `TombstonesConsumed`

WillemKauf committed:
`worker`: use `LatestValueProduced` in workers

As a means to verify the results for a compacted topic. Use the flag
`--validate-latest-values` to trigger validation.

WillemKauf committed:
`worker`: add `LatestValueMap`

WillemKauf committed:
`worker`: add `LatestValueProduced`

For verification of compacted topics, we can track the last
expected key-value pair that will be seen after the log has been
fully compacted.

WillemKauf committed:
`worker`: generate valid UTF-8 payloads

WillemKauf committed:
add `LatestConsumableOffsets` to `producer_worker`

And use it in the verifier workers.

In the case that the topic being consumed from had tombstones produced,
the high watermark may be given by a tombstone record that has been removed.
In trying to consume until this point, sequential readers will become stuck
polling for new records. Persist the last consumable offset in order to
adjust the offset we attempt to read up to in the read workers.

WillemKauf committed:
`producer_worker`: add `OnAcked()`

There was a race-y access to a map within the producer worker.
Add `OnAcked()` to allow for proper access of the producer's lock
during reading and writing.

WillemKauf committed:
Merge pull request #60 from WillemKauf/tombstones

Compaction verification enhancements

WillemKauf committed:
`verifier`: fix `lastConsumableOffsets`

This didn't account for tombstones being produced for the same key
later in the log.

Fix the logic by adding a new `latestKoByPartition` map which
tracks the last offset for a given key. By using it in combination
with `latestKvByPartition`, we can conclude which non-tombstone record/key
has the highest offset for a given partition.

WillemKauf committed:
Merge pull request #61 from WillemKauf/tombstones

`verifier`: fix `lastConsumableOffsets`

travisdowns committed:
kgo-repeater: listen for SIGTERM too

We listen for SIGINT and do a graceful shutdown, but we should do the
same thing on SIGTERM too, as that's what systemd will send a
service, and we run kgo as a service (and anyway it just makes sense
to do graceful shutdown on SIGTERM, that's kind of what it's there
for).

The main reason to do graceful shutdown is to commit any cached offsets
and (especially) to leave the consumer group explicitly, otherwise the
CG will stay in the Stable state until heartbeat timeout, which causes
problems for re-joining members.

travisdowns committed:
kgo-repeater: call Shutdown()

The verifiers expose a Shutdown() method but we never call it.

This means that the clients are never closed, so consumers never
explicitly leave their CG, which causes problems.

Fixes CORE-8470.

travisdowns committed:
Merge pull request #62 from redpanda-data/td-kgo-leave-group

Repeater should leave CG on graceful shutdown

WillemKauf committed:
`producer_worker`: revert use of `ValueGenerator::Generate()`

Using the modified `Generate()` function was a lot slower than just simply making
an empty payload:

```
Benchmark_random_payload10-32    2445903       485.2 ns/op
Benchmark_random_payload100-32    678870        1773 ns/op
Benchmark_random_payload1000-32   105667       13700 ns/op
Benchmark_random_payload1e4-32      9104      126782 ns/op
Benchmark_random_payload1e5-32      1135     1151425 ns/op
Benchmark_random_payload1e6-32       100    13046547 ns/op
==========================================================
Benchmark_empty_payload10-32    79950942       20.54 ns/op
Benchmark_empty_payload100-32   24089101       54.78 ns/op
Benchmark_empty_payload1000-32   3789248       366.1 ns/op
Benchmark_empty_payload1e4-32     670096        2059 ns/op
Benchmark_empty_payload1e5-32      58912       23021 ns/op
Benchmark_empty_payload1e6-32       5600      290045 ns/op
```

As can be seen, generating the random payload and ensuring it is UTF-8 valid
is orders of magnitude slower than the empty payload.

Revert the use of `Generate()` here in place of the empty payload.

However, if the user has indicated they want to validate the latest key-value
pair produced, generate a `(value-{%018d}, offset)` message in the record.

This does mean that message sizes that are less than 24 bytes are not honored
if the `validate-latest-values` flag is passed.

WillemKauf committed:
Merge pull request #63 from WillemKauf/tombstones

`producer_worker`: revert use of `ValueGenerator::Generate()`

nvartolomei committed:
Merge pull request #56 from redpanda-data/nv/crash-data-loss

crash instantly on data loss errors

dotnwat committed:
remove unused payload variable

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
produce random data in non-compressible mode

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
split out separate data generation tasks

this is useful to avoid needing to refactor other users of the generate
interface like kgo-repeater

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
add benchmakr for old and new random gen

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
use faster random data generator

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
add option for real random bytes

Signed-off-by: Noah Watkins <noahwatkins@gmail.com>

dotnwat committed:
Merge pull request #64 from dotnwat/random-data-produce

Random data produce

mmaslankaprv committed:
gomod: udpated franzgo version to the latest

Updated the version of franzgo clients as there were a lot of bugfixes
in franzgo in the rage of last 2 years.

Signed-off-by: Michał Maślanka <michal@redpanda.com>

mmaslankaprv committed:
Merge pull request #65 from mmaslankaprv/update-franz-go

gomod: udpated franzgo version to the latest

nvartolomei committed:
fix race condition when /status is polled before workers are registered

Also return an error on the status endpoint if workers are not fully
registered yet. Our callers expect that there is always at least one
worker report on the status endpoint.

nvartolomei committed:
Merge pull request #66 from redpanda-data/nv/fix-race

fix race condition when /status is polled before workers are registered

mmaslankaprv commented on tests/docker/ducktape-deps/kgo-verifier:
```diff
@@ -1,7 +1,5 @@
 cd /opt/kgo-verifier
-git reset --hard 319c69da3d7690e26d390261c4b331703b99260c
 go mod tidy
 make
```
use bazel here to build the verifier

nvartolomei replied:
i followed this from rp-storage-tool; we don't pull bazel for building these; should we? 🤔 i thought it would be more heavy weight but didn't validate that guess

nvartolomei commented:
/ci-repeat 1
skip-rebase

nvartolomei committed:
tests/go/kgo-verifier: update gitignore to ignore only the binary

nvartolomei commented:
/ci-repeat 1
skip-rebase
tests/rptest/tests/services_self_test.py

nvartolomei commented:
/ci-repeat 1
skip-rebase
tests/rptest/tests/services_self_test.py

rockwotj commented on tests/go/go-kafka-serde/client.go:
```diff
@@ -23,7 +23,7 @@ import (
 	log "github.com/sirupsen/logrus"
 
-	"github.com/redpanda-data/kgo-verifier/pkg/util"
+	"github.com/redpanda-data/redpanda/tools/kgo-verifier/pkg/util"
```
I think we should move `kgo-verifier` into `tests/go` and also hoist the `go.mod` to be `tests/go/go.mod` and share between all the versions. We really should be sharing these dependencies.

I could also be convinced to do this in a followup, but let's at least move `kgo-verifier` into the `tests/go`

nvartolomei replied:
I thought about this but couldn't convince myself to do it. I'm thinking about kgo-verifier as a generic tool mainly use for testing but it is also useful outside of our regular ducktape runs. I use it from cli/adhoc from time to time, etc.

Happy to move it as I only slightly lean towards `tools`. I also saw there other "testing" tools.

Single go.mod is a good argument for moving everything together. Though this won't solve in the general case all the problems as I can imagine dependencies on rpk, in theory. Other stuff might come up in the future too.

Less go mods is better so I'll probably move. Let me know if you didn't change your mind.

rockwotj replied:
I think it's fine to run stuff in testing manually as well 🤷 

> Single go.mod is a good argument for moving everything together. Though this won't solve in the general case all the problems as I can imagine dependencies on rpk, in theory. Other stuff might come up in the future too.

Yes agreed. If I had things my way we would share a `go.mod` with rpk, but since rpk isn't "owned" by core, it's not so simple.

nvartolomei replied:
Should be good now.

rockwotj replied:
Were we going to consolidate the `tests/go` `go.mod` too?

nvartolomei replied:
> Were we going to consolidate the tests/go go.mod too?

I'd make it in another PR since it will have to change other things beside kgo-verifier. Don't want to get side tracked before I reap the value of importing kgo.

nvartolomei committed:
Import https://github.com/redpanda-data/kgo-verifier

nvartolomei committed:
tests/go/kgo-verifier: update imports after importing repo

nvartolomei committed:
tests/go/kgo-verifier: bazelize

nvartolomei committed:
tests/docker: add new build script for kgo-verifier

Temporarily we will have 2 scripts to avoid the chicken-egg problem. Old
script is still used by CDT infrastructure. The new one is used by
dockerized ducktape tests.

The old one will be removed after we migrate CDT infra too.

nvartolomei committed:
tests/go/kgo-verifier: go mod tidy

nvartolomei commented:
/ci-repeat 1
skip-rebase
tests/rptest/tests/services_self_test.py

rockwotj commented on tests/go/kgo-verifier/.github/workflows/go.yml:
not needed?

rockwotj commented on tests/go/kgo-verifier/.github/dependabot.yml:
not needed?

rockwotj commented on tests/go/kgo-verifier/BSL.md:
not needed now?

rockwotj reviewed (dismissed):
LGTM, I think there are a few files we can nuke

nvartolomei committed:
tests/go/kgo-verifier: remove unecessary files post import

nvartolomei commented:
/ci-repeat 1
skip-rebase
tests/rptest/tests/services_self_test.py
